### ISSUE TITLE
does not work on webos 6

### ISSUE THREAD

**Issue by z31978** [PredatH0r/ChanSort#265]
does not work on webos 6

**Comment by droopy987** [comment]
Hi, 
I also would like to write an issue as I notice a strange behviour with webos 6. I don't know, what the problem of @z31978 is but I can explain my problem. 

I just bought a new LG C1 with webos 6. My current firmware version number is 03.15.95. I am using it with cable (cable from Telekom in Germany, pretty unusual but it is really cable from telekom ;-) ) and disabled the automatic channel update before exporting and importing. I also follow your guideline and ran a full channel search. Here is the first thing that I noticed. After the channel search, there is not a useless order, it is kind of presorted out of the box. For example DAS Erste is 1, ZDF is 2 and so on, even with disabled auto autopdate.

Then I exported the file to ChanSort and brought the channels in my favorite order. I then imported it to the TV and at first everything was fine. The channels were exactly ordered as I did in Chansort. After a short while (let's say 30 minutes or so) the order of the channel got suddenly reordered like the way it was after the channel search. Really strange. I then imported my ordered list again and again the channels were at first correctly ordered like it should be. Again after a while, they got reordered in the wrong order... 

That is very annoying me. 

At the moment, I don't have a solution for this. As a workaround I am importing my correct file again and again, when my channels get wrong again. But that is not a lasting solution. 

I also try it with the hints in this post [https://github.com/PredatH0r/ChanSort/discussions/207#discussioncomment-1066928](url), but when trying to import the file, my TV said, that it is not able to import the file...

So attached you find the channel list files. One directly after the full search and then my ordered list.

Can you maybe see, where the issue might be? I don't want to order my channel via the TV directly. That is the solution, LG suggested...

If you have any questions, just contact me! Thanks you! :-) 
[GlobalClone00001_original.zip](https://github.com/PredatH0r/ChanSort/files/7323222/GlobalClone00001_original.zip)
[GlobalClone00001_sortiert.zip](https://github.com/PredatH0r/ChanSort/files/7323223/GlobalClone00001_sortiert.zip)

**Comment by PredatH0r** [comment]
Thanks for your feedback!

Regarding the hint about sorting webOS 5 lists:
Personally I think it should not matter if using "Swap" or "Insert Before" or "Insert After", the resulting file will be the same - as long as channels are not deleted and no gaps in the numbers exist.

I only noticed a few weeks ago that my recommendation about a free channel search only affects the (Astra 19.2) satellite channel search. Because Astra 19.2 does not transmit any preferred channel numbers (LCN), the TV will start with an initial order based on frequency/transponder. When using a "provider" during the satellite search, the TV will use country specific tables built into its firmware to order the lists.

For DVB-C things are different, because most cable TV operators transmit LCN (logical channel numbers), which are used by the TV for the initial order. Normally, when disabling automatic update, you should be able to reorder the list as you see fit.
However, back in 2010 LG already had trouble with some models (using NetCast firmware) that kept changing the order despite disabled auto-update (in the normal menu). For these models the work-around was to activate hotel mode and disable channel updates through the hotel menu. I don't know if such a menu exists or how it can be access in webOS 5 and 6.
With webOS 3-4 everything looked fine (after some firmware updates).

The file formats for webOS 5 and 6 are identical and ChanSort doesn't (and can't) makes any distinction between them.
It wouldn't be the first nor the 2nd time that something is broken in LG's firmware and needs fixing. 
Like when they changed the file format from webOS 4 (XML) to webOS 5 (JSON) and removed data about favorite lists from the file.

...but too much information I guess :)

There are some data fields in the file ("userEditChNumber", "userSelCHNo", "userCustomize") in the file which I leave unchanged at the moment. They all default to "false".  It would be worth a try if setting any of them to "true" would force the TV to keep the number. The "validLCN" and "LCNPriority" fields looks like the file is not based on locked-in LCN numbers and should allow rearranging.

Before we start with experiments on the file content:
Is it possible for you to load the original file back into your TV (with its default order) and use the TV's menu to swap channels 1 and 2? If that can be done, could you send me an export of that list?
By comparing the original and the swapped list, maybe I can identify any indicator in the file for changed numbers.

Thanks and best regards,
Horst

**Comment by droopy987** [comment]
Hello Horst,
interesting background information. It's interesting how the structure of this is and how LG changed it during the different versions. 

When performing my channel search I did not select any provider. So I guess there might be another issue causing my problem. Maybe, there is really something broken with the firmware. I already contaced LG-Support and reported this, but they just say: "Order your channel directly via TV, then it will be fine. Not with a 3rd-party software as this causes incompatibilities..." But maybe they would not admit, that they might have problems with this... ;-)

> 
> There are some data fields in the file ("userEditChNumber", "userSelCHNo", "userCustomize") in the file which I leave unchanged at the moment. They all default to "false". It would be worth a try if setting any of them to "true" would force the TV to keep the number. The "validLCN" and "LCNPriority" fields looks like the file is not based on locked-in LCN numbers and should allow rearranging.
> 
> Before we start with experiments on the file content: Is it possible for you to load the original file back into your TV (with its default order) and use the TV's menu to swap channels 1 and 2? If that can be done, could you send me an export of that list? By comparing the original and the swapped list, maybe I can identify any indicator in the file for changed numbers.
> 
> Thanks and best regards, Horst

Your idea regarding the data fields is good. It's definetely worth a try.  I will do it this evening, loading the default to the tv and swap channels 1 and 2. Then I will send you both files again and you can compare. Maybe this helps.

In general I can also agree with @z31978. You did a great job with this software! :-) Thank you very very much for this. 

I will write you later and send the files. 

Regards :-)

**Comment by PredatH0r** [comment]
Great, thanks!

It would also be interesting if the TV keeps the new order stable after re-importing the list right after the export, or if it also starts shuffling that list around (in which case it would clearly proof an error in the firmware as there is no 3rd party editing involved)

**Comment by PredatH0r** [comment]
Thanks a lot for your files.
I compared them and indeed the "userEditChNumber" and "userSelCHNo" fields were set to "true".
Here is a new version of ChanSort that does the same: https://chansort.com/ChanSort_2021-10-12.zip

I took your original, swapped the 2 channels with this version and compared the result with your swapped, they were binary identical. I hope that works with your TV now.
There is a small chance that the import won't change any channels there were marked "edited" in the TV. That could be easily tested by loading your swapped first and then loading the original and verify if #1 is Das Erste and #2 is ZDF. If so, everything should be good to go. 
If #1 is still ZDF, then the edit-flag prevents changes by the import and a reset would be needed to import a list. (unlikely though)

**Comment by droopy987** [comment]
Hi again,
sorry, for the delay. I forgot to answer. I can tell you, that the issue is gone. The channels are still in the correct order. So, your change did this. Thank you very much!! :-)

**Comment by codefalk** [comment]
Hi there, I just bough a new LG OLED65G19LA OLED TV with WebOS 6.2 and try to import the tll file. I only get an error message. Is this a known restriction/bug?
Thanks
[GlobalClone00001.zip](https://github.com/PredatH0r/ChanSort/files/7739599/GlobalClone00001.zip)
